### PR TITLE
optimize group_norm op backward

### PR DESCRIPTION
**Pull request by Zjq9409**
### PR types

Performance optimization

### PR changes

OPs

### Describe

优化group_norm反向计算公式、添加向量化计算代码，反向计算性能数据如下：

op 配置 | 竞品耗时 | 优化前 | 优化前相比竞品 | 优化后 | 优化后相比竞品 | 加速比
-- | -- | -- | -- | -- | -- | --
shape:[2,256,38,26]   format:NCHW num_groups:32 | 0.02885 | 0.04573 | 差于 (58.51%) | 0.02180 | 优于 (24.70%) | 2.10
shape:[2,256,152,104]   format:NCHW num_groups:32 | 0.21768 | 0.22975 | 差于 (5.54%) | 0.19293 | 优于 (12.56%) | 1.19
shape:[2,256,76,52]   format:NCHW num_groups:32 | 0.07168 | 0.07838 | 差于 (9.35%) | 0.06992 | 打平 (3.97%) | 1.12
shape:[2,256,25,33]   format:NCHW num_groups:32 | 0.02506 | 0.03627 | 差于 (44.73%) | 0.01939 | 优于 (23.21%) | 1.87
shape:[2,256,100,132]   format:NCHW num_groups:32 | 0.18664 | 0.19681 | 差于 (5.45%) | 0.16327 | 优于 (10.37%) | 1.21
shape:[2,256,50,66]   format:NCHW num_groups:32 | 0.06226 | 0.07368 | 差于 (18.34%) | 0.06251 | 打平 (0.63%) | 1.18
shape:[2,256,10,11]   format:NCHW num_groups:32 | 0.01426 | 0.01283 | 优于 (10.03%) | 0.01133 | 优于 (21.16%) | 1.13
shape:[2,256,13,13]   format:NCHW num_groups:32 | 0.01529 | 0.01422 | 优于 (7.00%) | 0.01121 | 优于 (25.66%) | 1.27
shape:[2,256,144,128]   format:NCHW num_groups:32 | 0.25624 | 0.26543 | 打平 (3.59%) | 0.21781 | 优于 (13.19%) | 1.22
shape:[2,256,74,68]   format:NCHW num_groups:32 | 0.07861 | 0.09157 | 差于 (16.49%) | 0.07826 | 打平 (1.62%) | 1.17
shape:[2,128,25,42]   format:NCHW num_groups:32 | 0.01751 | 0.02577 | 差于 (47.17%) | 0.01512 | 优于 (10.85%) | 1.70
shape:[2,128,70,76]   format:NCHW num_groups:32 | 0.05347 | 0.05619 | 差于 (5.09%) | 0.04923 | 优于 (7.86%) | 1.14
shape:[2,256,280,304]   format:NCHW num_groups:32 | 1.10346 | 1.17337 | 差于 (6.34%) | 0.94262 | 优于 (14.38%) | 1.24
shape:[2,256,304,232]   format:NCHW num_groups:32 | 0.90551 | 0.97509 | 差于 (7.68%) | 0.77200 | 优于 (15.32%) | 1.26
shape:[2,128,35,38]   format:NCHW num_groups:32 | 0.01892 | 0.02555 | 差于 (35.04%) | 0.01650 | 优于 (9.89%) | 1.55
shape:[2,256,272,312]   format:NCHW num_groups:32 | 1.10848 | 1.17801 | 差于 (6.27%) | 0.93751 | 优于 (15.32%) | 1.26
shape:[2,256,320,312]   format:NCHW num_groups:32 | 1.28528 | 1.37059 | 差于 (6.64%) | 1.07318 | 优于 (17.14%) | 1.28
shape:[2,256,280,200]   format:NCHW num_groups:32 | 0.75756 | 0.80689 | 差于 (6.51%) | 0.64160 | 优于 (14.67%) | 1.26
shape:[2,128,140,152]   format:NCHW num_groups:32 | 0.15343 | 0.15773 | 打平 (2.80%) | 0.13413 | 优于 (12.52%) | 1.18
shape:[2,128,136,156]   format:NCHW num_groups:32 | 0.16582 | 0.16450 | 打平 (0.80%) | 0.13413 | 优于 (19.37%) | 1.23

前向+反向的性能数据如下：

op 配置 | 竞品耗时 | 优化前 | 优化前相比竞品 | 优化后 | 优化后相比竞品 | 加速比
-- | -- | -- | -- | -- | -- | --
shape:[2,256,38,26]   format:NCHW num_groups:32 | 0.05002 | 0.08798 | 差于 (75.89%) | 0.03881 | 优于 (22.61%) | 2.27
shape:[2,256,152,104]   format:NCHW num_groups:32 | 0.44987 | 0.39136 | 优于 (13.01%) | 0.34259 | 优于 (24.31%) | 1.14
shape:[2,256,76,52]   format:NCHW num_groups:32 | 0.14665 | 0.13921 | 优于 (5.07%) | 0.11571 | 优于 (21.81%) | 1.20
shape:[2,256,25,33]   format:NCHW num_groups:32 | 0.04453 | 0.07385 | 差于 (65.84%) | 0.03541 | 优于 (20.84%) | 2.09
shape:[2,256,100,132]   format:NCHW num_groups:32 | 0.38314 | 0.33415 | 优于 (12.79%) | 0.28951 | 优于 (24.59%) | 1.15
shape:[2,256,50,66]   format:NCHW num_groups:32 | 0.12724 | 0.13091 | 打平 (2.88%) | 0.10457 | 优于 (17.75%) | 1.25
shape:[2,256,10,11]   format:NCHW num_groups:32 | 0.02663 | 0.02523 | 优于 (5.26%) | 0.02228 | 优于 (16.68%) | 1.13
shape:[2,256,13,13]   format:NCHW num_groups:32 | 0.02831 | 0.02853 | 打平 (0.78%) | 0.02237 | 优于 (20.42%) | 1.28
shape:[2,256,144,128]   format:NCHW num_groups:32 | 0.51986 | 0.45103 | 优于 (13.24%) | 0.38964 | 优于 (24.38%) | 1.16
shape:[2,256,74,68]   format:NCHW num_groups:32 | 0.17181 | 0.16131 | 优于 (6.11%) | 0.13441 | 优于 (22.09%) | 1.20
shape:[2,128,25,42]   format:NCHW num_groups:32 | 0.03422 | 0.05010 | 差于 (46.41%) | 0.02949 | 优于 (12.54%) | 1.70
shape:[2,128,70,76]   format:NCHW num_groups:32 | 0.10730 | 0.09786 | 优于 (8.80%) | 0.08188 | 优于 (23.53%) | 1.20
shape:[2,256,280,304]   format:NCHW num_groups:32 | 2.25950 | 1.95465 | 优于 (13.49%) | 1.69268 | 优于 (24.68%) | 1.15
shape:[2,256,304,232]   format:NCHW num_groups:32 | 1.86614 | 1.62497 | 优于 (12.92%) | 1.39096 | 优于 (25.40%) | 1.17
shape:[2,128,35,38]   format:NCHW num_groups:32 | 0.03697 | 0.05034 | 差于 (36.16%) | 0.03142 | 优于 (13.56%) | 1.60
shape:[2,256,272,312]   format:NCHW num_groups:32 | 2.25328 | 1.97186 | 优于 (12.49%) | 1.69253 | 优于 (24.91%) | 1.17
shape:[2,256,320,312]   format:NCHW num_groups:32 | 2.63393 | 2.31147 | 优于 (12.24%) | 1.96908 | 优于 (25.41%) | 1.17
shape:[2,256,280,200]   format:NCHW num_groups:32 | 1.52122 | 1.34076 | 优于 (11.86%) | 1.14964 | 优于 (24.46%) | 1.17
shape:[2,128,140,152]   format:NCHW num_groups:32 | 0.31577 | 0.27354 | 优于 (13.37%) | 0.24205 | 优于 (23.43%) | 1.13
shape:[2,128,136,156]   format:NCHW num_groups:32 | 0.32905 | 0.27918 | 优于 (15.16%) | 0.24215 | 优于 (26.38%) | 1.15

### CODE REVIEW

File: paddle/fluid/operators/group_norm_op.cu [hunk 1]
```diff
@@ -29,6 +29,7 @@ namespace operators {
 
 using DataLayout = framework::DataLayout;
 enum GroupNormKernelFlags { kHasScale = 1, kHasBias = 2 };
+#define ALIGN_BYTES 16
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu:32]: 这个变量不是定义过了？

**Comment by ZzSean** [comment]
PR描述里简单说下优化方法

File: paddle/fluid/operators/group_norm_op.cu [hunk 2]
```diff
@@ -379,24 +376,180 @@ __global__ void GroupNormBackward(const T* x, const T* d_y, const T* scale,
...
+    T v_y = (tmp - x_bias) * x_scale_inv;
+    T dly = d_y[(bid * H + hid) * W * C + wid * C + ccid];
+    d_x[(bid * H + hid) * W * C + wid * C + ccid] =
+        x_var_inv *
+        (dly * x_scale - number_inv * d_x_var * v_y - number_inv * d_x_mean);
+  }
+}
+
+template <typename T, typename AccT, int VecSize>
+__device__ __forceinline__ void DThreadReduce(const T* x, const T* dy, int size,
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu]: 这函数是否可以跟前向的统一

File: paddle/fluid/operators/group_norm_op.cu [hunk 3]
```diff
@@ -433,21 +588,21 @@ class GroupNormGradKernel<platform::CUDADeviceContext, T>
...
-    T* temp_var_data = temp_var.data<T>();
-
-    Tensor temp_mean;
-    temp_mean.mutable_data<T>(var->dims(), ctx.GetPlace());
-    set_zero(dev_ctx, &temp_mean, static_cast<T>(0));
-    T* temp_mean_data = temp_mean.data<T>();
+    Tensor ds, db;
+    ds.mutable_data<T>({x_dims[0], C}, ctx.GetPlace());
+    db.mutable_data<T>({x_dims[0], C}, ctx.GetPlace());
+    set_zero(dev_ctx, &ds, static_cast<T>(0));
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu]: 这里不是还是会有set_zero吗

File: paddle/fluid/operators/group_norm_op.cu [hunk 4]
```diff
@@ -486,15 +641,76 @@ class GroupNormGradKernel<platform::CUDADeviceContext, T>
...
+      if (d_scale || d_bias) {
+        const int block = 256;
+        GetScaleBiasGradientCUDAKernel<
+            T><<<(C + block - 1) / block, block, 0, dev_ctx.stream()>>>(
+            x_dims[0], C, groups, epsilon, mean_data, var_data, ds_data,
+            db_data, d_scale_data, d_bias_data);
+      }
+
+      if (d_x_data != nullptr) {
+        Tensor p1, p2, p3;
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu:674]: 这些变量名都是什么意义？

**Comment by Zjq9409** [comment]
> PR描述里简单说下优化方法

已加

File: paddle/fluid/operators/group_norm_op.cu (hunk 3 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu]: 已经删除

File: paddle/fluid/operators/group_norm_op.cu (hunk 4 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu:674]: 是计算反向公式使用的临时变量

File: paddle/fluid/operators/group_norm_op.cu (hunk 2 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu]: Done.

File: paddle/fluid/operators/group_norm_op.cu [hunk 5]
```diff
@@ -81,46 +81,66 @@ __global__ void GroupNormForwardGetMeanAndVar(const T* x, int N, int C, int W,
...
-__device__ __forceinline__ void ThreadReduce(const T* input, int size,
-                                             const int offset, AccT* mean,
-                                             AccT* var) {
+template <typename T, typename AccT, int VecSize, int Num>
+__device__ __forceinline__ void ThreadReduce(phi::Array<const T*, Num> arrs,
+                                             int size, const int offset,
+                                             AccT* out_mean, AccT* out_var) {
+  const T* x = arrs[0];
+  const T* y;
+  if (Num == 2) y = arrs[1];
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu]: if 语句使用大括号，下同

File: paddle/fluid/operators/group_norm_op.cu [hunk 6]
```diff
@@ -148,7 +168,10 @@ __global__ void VectorizedGetMeanAndVarNCHW(const T* x, T* mean, T* var,
   AccT x_var = static_cast<AccT>(0);
   const int input_offset = ((uint64_t)x) % ALIGN_BYTES / sizeof(T);
   x += i * size;
-  ThreadReduce<T, AccT, VecSize>(x, size, input_offset, &x_mean, &x_var);
+  phi::Array<const T*, 1> ins;
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu:171]: 你这个ins初始化长度只有1，不太对吧

File: paddle/fluid/operators/group_norm_op.cu (hunk 4 above)
Comment (ZzSean) [reply paddle/fluid/operators/group_norm_op.cu:674]: 我的意思是这些变量让别人看起来根本不知道啥含义，你这里可以换个名字或者注释一下

File: paddle/fluid/operators/group_norm_op.cu (hunk 5 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu]: Done.

File: paddle/fluid/operators/group_norm_op.cu (hunk 4 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu:674]: Done.

File: paddle/fluid/operators/group_norm_op.cu [hunk 7]
```diff
@@ -379,26 +407,142 @@ __global__ void GroupNormBackward(const T* x, const T* d_y, const T* scale,
...
+    const T x = (sum2 * static_cast<T>(mean[ng]) - sum1) *
+                static_cast<T>(var_inv) * static_cast<T>(var_inv) *
+                static_cast<T>(var_inv) * s;
+    p2[ng] = x;
+    p3[ng] = -x * static_cast<T>(mean[ng]) - sum2 * static_cast<T>(var_inv) * s;
+  }
+}
+
+template <typename T>
+__global__ void GetDxGradientCUDAKernel(int imsize, int C, int group_size,
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu]: 这个函数命名，Dx本来就表示x的grad了，后面再接个Gradient不太合适，后续PR再修改吧

File: paddle/fluid/operators/group_norm_op.cu [hunk 8]
```diff
@@ -479,22 +621,93 @@ class GroupNormGradKernel<platform::CUDADeviceContext, T>
...
+            dim3(x_dims[0], groups), block_dims, 0, dev_ctx.stream()>>>(
+            imsize, groups, group_size, epsilon, mean_data, var_data,
+            scale_data, ds_data, db_data, p1_data, p2_data, p3_data);
+        GetDxGradientCUDAKernel<T><<<grid, threads, 0, dev_ctx.stream()>>>(
+            imsize, C, group_size, groups, p1_data, p2_data, p3_data, x_data,
+            dy_data, d_x_data);
+      }
+
+    } else {
+      set_zero(dev_ctx, d_scale, static_cast<T>(0));
```
Comment (ZzSean) [inline paddle/fluid/operators/group_norm_op.cu]: 这里不用判断d_scale和d_bias是否为空吗？会不会有问题？

File: paddle/fluid/operators/group_norm_op.cu (hunk 7 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu]: Done. 已经修改函数名称

File: paddle/fluid/operators/group_norm_op.cu (hunk 8 above)
Comment (Zjq9409) [reply paddle/fluid/operators/group_norm_op.cu]: 加上了是否为空的判断

**Review by ZzSean — APPROVED** [review]
LGTM

**Pull request merged**